Mocking Vault Provider Fails with Missing 'address' Attribute in OpenTofu v1.8.2 in test command and fails to retrieve the secret

### OpenTofu Version

```shell
OpenTofu v1.8.2
on darwin_arm64
+ provider registry.terraform.io/hashicorp/vault v4.4.0
```


### OpenTofu Configuration Files

main.tf
```hcl
# Define Variables
variable "customer_identifier" {
  type = string
}

variable "account_identifier" {
  type = string
}

variable "environment_identifier" {
  type = string
}

variable "instance_identifier" {
  type = string
}

variable "vault_secret_path" {
  type = string
}

# Define Vault provider configuration
provider "vault" {
  alias   = "mock"
  address = "http://127.0.0.1:8200"
}

# Data block to access a secret in Vault
data "vault_kv_secret_v2" "self" {
  name  = var.vault_secret_path
  mount = "kv"
}

# Output the retrieved secret value for debugging purposes
output "vault_secret_data" {
  value     = data.vault_kv_secret_v2.self.data
  sensitive = true
}
```

vault.tftest.hcl
```hcl
# Define Variables
variables {
  customer_identifier    = "cust123"
  account_identifier     = "acc456"
  environment_identifier = "env789"
  instance_identifier    = "inst101"
  vault_secret_path      = "test/acc456/env789/secret"
}


# provider "vault" {
#   alias   = "mock"
#   address = "http://127.0.0.1:8200"
# }

mock_provider "vault" {
  alias = "mock"

  mock_data "vault_kv_secret_v2" {
    defaults = {
      data = {
        "test" = "mocked-password"
      }
    }
  }
}

run "validate_vault_secret" {
  command = apply
  providers = {
    vault = vault.mock
  }

  # Validate the stored admin password
  assert {
    condition     = data.vault_kv_secret_v2.self.data["test"] == "mocked-password"
    error_message = "Error quering the Vault Secret - Super user password"
  }

}
```

### Debug Output

**When `address` is not mentioned in the non-test provider**

```log
tofu init; tofu test -filter vault.tftest.hcl -verbose
2024-11-05T03:09:22.565+0200 [INFO]  OpenTofu version: 1.8.2
2024-11-05T03:09:22.566+0200 [DEBUG] using github.com/hashicorp/go-tfe v1.36.0
2024-11-05T03:09:22.566+0200 [DEBUG] using github.com/opentofu/hcl/v2 v2.0.0-20240416130056-03228b26f391
2024-11-05T03:09:22.566+0200 [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.1
2024-11-05T03:09:22.566+0200 [DEBUG] using github.com/zclconf/go-cty v1.14.4
2024-11-05T03:09:22.566+0200 [INFO]  Go runtime version: go1.23.0
2024-11-05T03:09:22.566+0200 [INFO]  CLI args: []string{"tofu", "init"}
2024-11-05T03:09:22.566+0200 [DEBUG] Attempting to open CLI config file: /Users/jishnu.sygal/.tofurc
2024-11-05T03:09:22.566+0200 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2024-11-05T03:09:22.567+0200 [DEBUG] checking for credentials in "/Users/jishnu.sygal/.terraform.d/plugins"
2024-11-05T03:09:22.567+0200 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2024-11-05T03:09:22.567+0200 [DEBUG] will search for provider plugins in /Users/jishnu.sygal/.terraform.d/plugins
2024-11-05T03:09:22.567+0200 [DEBUG] ignoring non-existing provider search directory /Users/jishnu.sygal/Library/Application Support/io.terraform/plugins
2024-11-05T03:09:22.567+0200 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2024-11-05T03:09:22.568+0200 [DEBUG] Found the config directory: /Users/jishnu.sygal/.terraform.d
2024-11-05T03:09:22.568+0200 [INFO]  CLI command args: []string{"init"}

Initializing the backend...
2024-11-05T03:09:22.573+0200 [DEBUG] New state was assigned lineage "5732cf7d-839a-fb61-f13c-9436ab5a7629"
2024-11-05T03:09:22.606+0200 [DEBUG] checking for provisioner in "."
2024-11-05T03:09:22.606+0200 [DEBUG] checking for provisioner in "/opt/homebrew/bin"
2024-11-05T03:09:22.606+0200 [DEBUG] checking for provisioner in "/Users/jishnu.sygal/.terraform.d/plugins"

Initializing provider plugins...
- Reusing previous version of hashicorp/vault from the dependency lock file
2024-11-05T03:09:22.608+0200 [DEBUG] Service discovery for registry.opentofu.org at https://registry.opentofu.org/.well-known/terraform.json
2024-11-05T03:09:22.731+0200 [DEBUG] GET https://registry.opentofu.org/v1/providers/hashicorp/vault/versions
- Using previously-installed hashicorp/vault v4.4.0

OpenTofu has been successfully initialized!

You may now begin working with OpenTofu. Try running "tofu plan" to see
any changes that are required for your infrastructure. All OpenTofu commands
should now work.

If you ever set or change modules or backend configuration for OpenTofu,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
2024-11-05T03:09:22.833+0200 [INFO]  OpenTofu version: 1.8.2
2024-11-05T03:09:22.833+0200 [DEBUG] using github.com/hashicorp/go-tfe v1.36.0
2024-11-05T03:09:22.833+0200 [DEBUG] using github.com/opentofu/hcl/v2 v2.0.0-20240416130056-03228b26f391
2024-11-05T03:09:22.833+0200 [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.1
2024-11-05T03:09:22.833+0200 [DEBUG] using github.com/zclconf/go-cty v1.14.4
2024-11-05T03:09:22.833+0200 [INFO]  Go runtime version: go1.23.0
2024-11-05T03:09:22.833+0200 [INFO]  CLI args: []string{"tofu", "test", "-filter", "vault.tftest.hcl", "-verbose"}
2024-11-05T03:09:22.833+0200 [DEBUG] Attempting to open CLI config file: /Users/jishnu.sygal/.tofurc
2024-11-05T03:09:22.833+0200 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2024-11-05T03:09:22.833+0200 [DEBUG] checking for credentials in "/Users/jishnu.sygal/.terraform.d/plugins"
2024-11-05T03:09:22.833+0200 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2024-11-05T03:09:22.833+0200 [DEBUG] will search for provider plugins in /Users/jishnu.sygal/.terraform.d/plugins
2024-11-05T03:09:22.834+0200 [DEBUG] ignoring non-existing provider search directory /Users/jishnu.sygal/Library/Application Support/io.terraform/plugins
2024-11-05T03:09:22.834+0200 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2024-11-05T03:09:22.834+0200 [DEBUG] Found the config directory: /Users/jishnu.sygal/.terraform.d
2024-11-05T03:09:22.834+0200 [INFO]  CLI command args: []string{"test", "-filter", "vault.tftest.hcl", "-verbose"}
2024-11-05T03:09:22.835+0200 [DEBUG] TestCommand: found 1 files with 1 run blocks
2024-11-05T03:09:22.853+0200 [DEBUG] checking for provisioner in "."
2024-11-05T03:09:22.854+0200 [DEBUG] checking for provisioner in "/opt/homebrew/bin"
2024-11-05T03:09:22.854+0200 [DEBUG] checking for provisioner in "/Users/jishnu.sygal/.terraform.d/plugins"
2024-11-05T03:09:22.858+0200 [DEBUG] TestFileRunner: starting validate for vault.tftest.hcl/validate_vault_secret
2024-11-05T03:09:22.859+0200 [DEBUG] Building and walking validate graph
2024-11-05T03:09:22.862+0200 [DEBUG] ProviderTransformer: "data.vault_kv_secret_v2.self" (*tofu.NodeValidatableResource) needs provider["registry.opentofu.org/hashicorp/vault"]
2024-11-05T03:09:22.862+0200 [DEBUG] created provider logger: level=debug
2024-11-05T03:09:22.862+0200 [INFO]  provider: configuring client automatic mTLS
2024-11-05T03:09:22.870+0200 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault args=[".terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault"]
2024-11-05T03:09:22.872+0200 [DEBUG] provider: plugin started: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault pid=53459
2024-11-05T03:09:22.872+0200 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault
2024-11-05T03:09:22.895+0200 [INFO]  provider.terraform-provider-vault: configuring server automatic mTLS: timestamp="2024-11-05T03:09:22.894+0200"
2024-11-05T03:09:22.901+0200 [DEBUG] provider: using plugin: version=5
2024-11-05T03:09:22.902+0200 [DEBUG] provider.terraform-provider-vault: plugin address: address=/var/folders/n4/5f364y013pv4gmq2gk64rntm0000gq/T/plugin2139298573 network=unix timestamp="2024-11-05T03:09:22.901+0200"
2024-11-05T03:09:22.922+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-11-05T03:09:22.923+0200 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault pid=53459
2024-11-05T03:09:22.923+0200 [DEBUG] provider: plugin exited
2024-11-05T03:09:22.924+0200 [DEBUG] pruning unused provider["registry.opentofu.org/hashicorp/vault"].mock
2024-11-05T03:09:22.924+0200 [DEBUG] ReferenceTransformer: "var.account_identifier" references: []
2024-11-05T03:09:22.924+0200 [DEBUG] ReferenceTransformer: "var.environment_identifier" references: []
2024-11-05T03:09:22.924+0200 [DEBUG] ReferenceTransformer: "var.instance_identifier" references: []
2024-11-05T03:09:22.924+0200 [DEBUG] ReferenceTransformer: "var.vault_secret_path" references: []
2024-11-05T03:09:22.924+0200 [DEBUG] ReferenceTransformer: "output.vault_secret_data (expand)" references: [data.vault_kv_secret_v2.self]
2024-11-05T03:09:22.924+0200 [DEBUG] ReferenceTransformer: "provider[\"registry.opentofu.org/hashicorp/vault\"]" references: []
2024-11-05T03:09:22.924+0200 [DEBUG] ReferenceTransformer: "data.vault_kv_secret_v2.self" references: [var.vault_secret_path]
2024-11-05T03:09:22.924+0200 [DEBUG] ReferenceTransformer: "var.customer_identifier" references: []
2024-11-05T03:09:22.925+0200 [DEBUG] Starting graph walk: walkValidate
2024-11-05T03:09:22.927+0200 [DEBUG] created provider logger: level=debug
2024-11-05T03:09:22.927+0200 [INFO]  provider: configuring client automatic mTLS
2024-11-05T03:09:22.930+0200 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault args=[".terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault"]
2024-11-05T03:09:22.932+0200 [DEBUG] provider: plugin started: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault pid=53462
2024-11-05T03:09:22.932+0200 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault
2024-11-05T03:09:22.944+0200 [INFO]  provider.terraform-provider-vault: configuring server automatic mTLS: timestamp="2024-11-05T03:09:22.943+0200"
2024-11-05T03:09:22.950+0200 [DEBUG] provider.terraform-provider-vault: plugin address: address=/var/folders/n4/5f364y013pv4gmq2gk64rntm0000gq/T/plugin1191921133 network=unix timestamp="2024-11-05T03:09:22.950+0200"
2024-11-05T03:09:22.950+0200 [DEBUG] provider: using plugin: version=5
2024-11-05T03:09:22.958+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-11-05T03:09:22.959+0200 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault pid=53462
2024-11-05T03:09:22.959+0200 [DEBUG] provider: plugin exited
2024-11-05T03:09:22.959+0200 [DEBUG] TestFileRunner: completed validate for  vault.tftest.hcl/validate_vault_secret
2024-11-05T03:09:22.959+0200 [DEBUG] TestFileRunner: starting plan for vault.tftest.hcl/validate_vault_secret
2024-11-05T03:09:22.960+0200 [DEBUG] Building and walking plan graph for NormalMode
2024-11-05T03:09:22.961+0200 [DEBUG] ProviderTransformer: "data.vault_kv_secret_v2.self (expand)" (*tofu.nodeExpandPlannableResource) needs provider["registry.opentofu.org/hashicorp/vault"]
2024-11-05T03:09:22.961+0200 [DEBUG] pruning unused provider["registry.opentofu.org/hashicorp/vault"].mock
2024-11-05T03:09:22.961+0200 [DEBUG] ReferenceTransformer: "var.environment_identifier" references: []
2024-11-05T03:09:22.961+0200 [DEBUG] ReferenceTransformer: "var.vault_secret_path" references: []
2024-11-05T03:09:22.961+0200 [DEBUG] ReferenceTransformer: "output.vault_secret_data (expand)" references: [data.vault_kv_secret_v2.self (expand)]
2024-11-05T03:09:22.961+0200 [DEBUG] ReferenceTransformer: "var.account_identifier" references: []
2024-11-05T03:09:22.961+0200 [DEBUG] ReferenceTransformer: "var.instance_identifier" references: []
2024-11-05T03:09:22.961+0200 [DEBUG] ReferenceTransformer: "var.customer_identifier" references: []
2024-11-05T03:09:22.961+0200 [DEBUG] ReferenceTransformer: "&{[0x14000614690]}" references: [data.vault_kv_secret_v2.self (expand)]
2024-11-05T03:09:22.961+0200 [DEBUG] ReferenceTransformer: "provider[\"registry.opentofu.org/hashicorp/vault\"]" references: []
2024-11-05T03:09:22.961+0200 [DEBUG] ReferenceTransformer: "data.vault_kv_secret_v2.self (expand)" references: [var.vault_secret_path]
2024-11-05T03:09:22.961+0200 [DEBUG] Starting graph walk: walkPlan
2024-11-05T03:09:22.961+0200 [DEBUG] created provider logger: level=debug
2024-11-05T03:09:22.961+0200 [INFO]  provider: configuring client automatic mTLS
2024-11-05T03:09:22.964+0200 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault args=[".terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault"]
2024-11-05T03:09:22.965+0200 [DEBUG] provider: plugin started: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault pid=53463
2024-11-05T03:09:22.966+0200 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault
2024-11-05T03:09:22.978+0200 [INFO]  provider.terraform-provider-vault: configuring server automatic mTLS: timestamp="2024-11-05T03:09:22.978+0200"
2024-11-05T03:09:22.984+0200 [DEBUG] provider: using plugin: version=5
2024-11-05T03:09:22.984+0200 [DEBUG] provider.terraform-provider-vault: plugin address: address=/var/folders/n4/5f364y013pv4gmq2gk64rntm0000gq/T/plugin2305479999 network=unix timestamp="2024-11-05T03:09:22.984+0200"
2024-11-05T03:09:22.989+0200 [ERROR] vertex "provider[\"registry.opentofu.org/hashicorp/vault\"]" error: Missing required attribute
2024-11-05T03:09:22.989+0200 [DEBUG] TestFileRunner: completed plan for vault.tftest.hcl/validate_vault_secret
vault.tftest.hcl... fail
  run "validate_vault_secret"... fail
╷
│ Error: Missing required attribute
│
│   on provider["registry.opentofu.org/hashicorp/vault"] with no configuration line 1:
│   (source code not available)
│
│ The attribute "address" is required, but no definition was found.
╵

Failure! 0 passed, 1 failed.
2024-11-05T03:09:22.990+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-11-05T03:09:22.990+0200 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault pid=53463
2024-11-05T03:09:22.990+0200 [DEBUG] provider: plugin exited
```



**When `address`  is provided in the non-test provider block**

```log
tofu init; tofu test -filter vault.tftest.hcl -verbose
2024-11-05T03:14:31.367+0200 [INFO]  OpenTofu version: 1.8.2
2024-11-05T03:14:31.367+0200 [DEBUG] using github.com/hashicorp/go-tfe v1.36.0
2024-11-05T03:14:31.367+0200 [DEBUG] using github.com/opentofu/hcl/v2 v2.0.0-20240416130056-03228b26f391
2024-11-05T03:14:31.367+0200 [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.1
2024-11-05T03:14:31.367+0200 [DEBUG] using github.com/zclconf/go-cty v1.14.4
2024-11-05T03:14:31.367+0200 [INFO]  Go runtime version: go1.23.0
2024-11-05T03:14:31.367+0200 [INFO]  CLI args: []string{"tofu", "init"}
2024-11-05T03:14:31.368+0200 [DEBUG] Attempting to open CLI config file: /Users/jishnu.sygal/.tofurc
2024-11-05T03:14:31.368+0200 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2024-11-05T03:14:31.368+0200 [DEBUG] checking for credentials in "/Users/jishnu.sygal/.terraform.d/plugins"
2024-11-05T03:14:31.368+0200 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2024-11-05T03:14:31.368+0200 [DEBUG] will search for provider plugins in /Users/jishnu.sygal/.terraform.d/plugins
2024-11-05T03:14:31.368+0200 [DEBUG] ignoring non-existing provider search directory /Users/jishnu.sygal/Library/Application Support/io.terraform/plugins
2024-11-05T03:14:31.368+0200 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2024-11-05T03:14:31.368+0200 [DEBUG] Found the config directory: /Users/jishnu.sygal/.terraform.d
2024-11-05T03:14:31.368+0200 [INFO]  CLI command args: []string{"init"}

Initializing the backend...
2024-11-05T03:14:31.369+0200 [DEBUG] New state was assigned lineage "f0db06e7-c2a0-e91c-fe17-65266b087041"
2024-11-05T03:14:31.404+0200 [DEBUG] checking for provisioner in "."
2024-11-05T03:14:31.405+0200 [DEBUG] checking for provisioner in "/opt/homebrew/bin"
2024-11-05T03:14:31.405+0200 [DEBUG] checking for provisioner in "/Users/jishnu.sygal/.terraform.d/plugins"

Initializing provider plugins...
- Reusing previous version of hashicorp/vault from the dependency lock file
2024-11-05T03:14:31.406+0200 [DEBUG] Service discovery for registry.opentofu.org at https://registry.opentofu.org/.well-known/terraform.json
2024-11-05T03:14:31.468+0200 [DEBUG] GET https://registry.opentofu.org/v1/providers/hashicorp/vault/versions
- Using previously-installed hashicorp/vault v4.4.0

OpenTofu has been successfully initialized!

You may now begin working with OpenTofu. Try running "tofu plan" to see
any changes that are required for your infrastructure. All OpenTofu commands
should now work.

If you ever set or change modules or backend configuration for OpenTofu,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
2024-11-05T03:14:31.563+0200 [INFO]  OpenTofu version: 1.8.2
2024-11-05T03:14:31.563+0200 [DEBUG] using github.com/hashicorp/go-tfe v1.36.0
2024-11-05T03:14:31.563+0200 [DEBUG] using github.com/opentofu/hcl/v2 v2.0.0-20240416130056-03228b26f391
2024-11-05T03:14:31.563+0200 [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.1
2024-11-05T03:14:31.563+0200 [DEBUG] using github.com/zclconf/go-cty v1.14.4
2024-11-05T03:14:31.563+0200 [INFO]  Go runtime version: go1.23.0
2024-11-05T03:14:31.563+0200 [INFO]  CLI args: []string{"tofu", "test", "-filter", "vault.tftest.hcl", "-verbose"}
2024-11-05T03:14:31.563+0200 [DEBUG] Attempting to open CLI config file: /Users/jishnu.sygal/.tofurc
2024-11-05T03:14:31.563+0200 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2024-11-05T03:14:31.564+0200 [DEBUG] checking for credentials in "/Users/jishnu.sygal/.terraform.d/plugins"
2024-11-05T03:14:31.564+0200 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2024-11-05T03:14:31.564+0200 [DEBUG] will search for provider plugins in /Users/jishnu.sygal/.terraform.d/plugins
2024-11-05T03:14:31.564+0200 [DEBUG] ignoring non-existing provider search directory /Users/jishnu.sygal/Library/Application Support/io.terraform/plugins
2024-11-05T03:14:31.564+0200 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2024-11-05T03:14:31.564+0200 [DEBUG] Found the config directory: /Users/jishnu.sygal/.terraform.d
2024-11-05T03:14:31.564+0200 [INFO]  CLI command args: []string{"test", "-filter", "vault.tftest.hcl", "-verbose"}
2024-11-05T03:14:31.564+0200 [DEBUG] TestCommand: found 1 files with 1 run blocks
2024-11-05T03:14:31.583+0200 [DEBUG] checking for provisioner in "."
2024-11-05T03:14:31.583+0200 [DEBUG] checking for provisioner in "/opt/homebrew/bin"
2024-11-05T03:14:31.584+0200 [DEBUG] checking for provisioner in "/Users/jishnu.sygal/.terraform.d/plugins"
2024-11-05T03:14:31.584+0200 [DEBUG] TestFileRunner: starting validate for vault.tftest.hcl/validate_vault_secret
2024-11-05T03:14:31.584+0200 [DEBUG] Building and walking validate graph
2024-11-05T03:14:31.584+0200 [DEBUG] ProviderTransformer: "data.vault_kv_secret_v2.self" (*tofu.NodeValidatableResource) needs provider["registry.opentofu.org/hashicorp/vault"]
2024-11-05T03:14:31.584+0200 [DEBUG] created provider logger: level=debug
2024-11-05T03:14:31.584+0200 [INFO]  provider: configuring client automatic mTLS
2024-11-05T03:14:31.590+0200 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault args=[".terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault"]
2024-11-05T03:14:31.592+0200 [DEBUG] provider: plugin started: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault pid=54195
2024-11-05T03:14:31.592+0200 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault
2024-11-05T03:14:31.610+0200 [INFO]  provider.terraform-provider-vault: configuring server automatic mTLS: timestamp="2024-11-05T03:14:31.610+0200"
2024-11-05T03:14:31.616+0200 [DEBUG] provider: using plugin: version=5
2024-11-05T03:14:31.616+0200 [DEBUG] provider.terraform-provider-vault: plugin address: address=/var/folders/n4/5f364y013pv4gmq2gk64rntm0000gq/T/plugin4047822842 network=unix timestamp="2024-11-05T03:14:31.616+0200"
2024-11-05T03:14:31.631+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-11-05T03:14:31.632+0200 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault pid=54195
2024-11-05T03:14:31.633+0200 [DEBUG] provider: plugin exited
2024-11-05T03:14:31.633+0200 [DEBUG] pruning unused provider["registry.opentofu.org/hashicorp/vault"].mock
2024-11-05T03:14:31.633+0200 [DEBUG] ReferenceTransformer: "data.vault_kv_secret_v2.self" references: [var.vault_secret_path]
2024-11-05T03:14:31.633+0200 [DEBUG] ReferenceTransformer: "var.customer_identifier" references: []
2024-11-05T03:14:31.633+0200 [DEBUG] ReferenceTransformer: "var.environment_identifier" references: []
2024-11-05T03:14:31.633+0200 [DEBUG] ReferenceTransformer: "var.vault_secret_path" references: []
2024-11-05T03:14:31.633+0200 [DEBUG] ReferenceTransformer: "var.account_identifier" references: []
2024-11-05T03:14:31.633+0200 [DEBUG] ReferenceTransformer: "var.instance_identifier" references: []
2024-11-05T03:14:31.633+0200 [DEBUG] ReferenceTransformer: "output.vault_secret_data (expand)" references: [data.vault_kv_secret_v2.self]
2024-11-05T03:14:31.633+0200 [DEBUG] ReferenceTransformer: "provider[\"registry.opentofu.org/hashicorp/vault\"]" references: []
2024-11-05T03:14:31.633+0200 [DEBUG] Starting graph walk: walkValidate
2024-11-05T03:14:31.633+0200 [DEBUG] created provider logger: level=debug
2024-11-05T03:14:31.633+0200 [INFO]  provider: configuring client automatic mTLS
2024-11-05T03:14:31.636+0200 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault args=[".terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault"]
2024-11-05T03:14:31.637+0200 [DEBUG] provider: plugin started: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault pid=54196
2024-11-05T03:14:31.638+0200 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault
2024-11-05T03:14:31.649+0200 [INFO]  provider.terraform-provider-vault: configuring server automatic mTLS: timestamp="2024-11-05T03:14:31.649+0200"
2024-11-05T03:14:31.655+0200 [DEBUG] provider: using plugin: version=5
2024-11-05T03:14:31.655+0200 [DEBUG] provider.terraform-provider-vault: plugin address: network=unix address=/var/folders/n4/5f364y013pv4gmq2gk64rntm0000gq/T/plugin452788671 timestamp="2024-11-05T03:14:31.655+0200"
2024-11-05T03:14:31.664+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-11-05T03:14:31.668+0200 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault pid=54196
2024-11-05T03:14:31.668+0200 [DEBUG] provider: plugin exited
2024-11-05T03:14:31.668+0200 [DEBUG] TestFileRunner: completed validate for  vault.tftest.hcl/validate_vault_secret
2024-11-05T03:14:31.668+0200 [DEBUG] TestFileRunner: starting plan for vault.tftest.hcl/validate_vault_secret
2024-11-05T03:14:31.668+0200 [DEBUG] Building and walking plan graph for NormalMode
2024-11-05T03:14:31.668+0200 [DEBUG] ProviderTransformer: "data.vault_kv_secret_v2.self (expand)" (*tofu.nodeExpandPlannableResource) needs provider["registry.opentofu.org/hashicorp/vault"]
2024-11-05T03:14:31.669+0200 [DEBUG] pruning unused provider["registry.opentofu.org/hashicorp/vault"].mock
2024-11-05T03:14:31.669+0200 [DEBUG] ReferenceTransformer: "data.vault_kv_secret_v2.self (expand)" references: [var.vault_secret_path]
2024-11-05T03:14:31.669+0200 [DEBUG] ReferenceTransformer: "var.customer_identifier" references: []
2024-11-05T03:14:31.669+0200 [DEBUG] ReferenceTransformer: "var.account_identifier" references: []
2024-11-05T03:14:31.669+0200 [DEBUG] ReferenceTransformer: "var.environment_identifier" references: []
2024-11-05T03:14:31.669+0200 [DEBUG] ReferenceTransformer: "var.vault_secret_path" references: []
2024-11-05T03:14:31.669+0200 [DEBUG] ReferenceTransformer: "var.instance_identifier" references: []
2024-11-05T03:14:31.669+0200 [DEBUG] ReferenceTransformer: "output.vault_secret_data (expand)" references: [data.vault_kv_secret_v2.self (expand)]
2024-11-05T03:14:31.669+0200 [DEBUG] ReferenceTransformer: "provider[\"registry.opentofu.org/hashicorp/vault\"]" references: []
2024-11-05T03:14:31.669+0200 [DEBUG] ReferenceTransformer: "&{[0x140010d93b0]}" references: [data.vault_kv_secret_v2.self (expand)]
2024-11-05T03:14:31.669+0200 [DEBUG] Starting graph walk: walkPlan
2024-11-05T03:14:31.669+0200 [DEBUG] created provider logger: level=debug
2024-11-05T03:14:31.669+0200 [INFO]  provider: configuring client automatic mTLS
2024-11-05T03:14:31.672+0200 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault args=[".terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault"]
2024-11-05T03:14:31.673+0200 [DEBUG] provider: plugin started: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault pid=54197
2024-11-05T03:14:31.674+0200 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault
2024-11-05T03:14:31.685+0200 [INFO]  provider.terraform-provider-vault: configuring server automatic mTLS: timestamp="2024-11-05T03:14:31.685+0200"
2024-11-05T03:14:31.691+0200 [DEBUG] provider: using plugin: version=5
2024-11-05T03:14:31.692+0200 [DEBUG] provider.terraform-provider-vault: plugin address: address=/var/folders/n4/5f364y013pv4gmq2gk64rntm0000gq/T/plugin4276247703 network=unix timestamp="2024-11-05T03:14:31.691+0200"
2024-11-05T03:14:31.699+0200 [WARN]  ValidateProviderConfig from "provider[\"registry.opentofu.org/hashicorp/vault\"]" changed the config value, but that value is unused
2024-11-05T03:14:31.699+0200 [DEBUG] Resource instance state not found for node "data.vault_kv_secret_v2.self", instance data.vault_kv_secret_v2.self
2024-11-05T03:14:31.699+0200 [INFO]  ReferenceTransformer: reference not found: "var.vault_secret_path"
2024-11-05T03:14:31.699+0200 [DEBUG] ReferenceTransformer: "data.vault_kv_secret_v2.self" references: []
2024-11-05T03:14:31.701+0200 [DEBUG] provider.terraform-provider-vault: [eb732054-0d99-45f1-99cd-9b0439a9a9a2] Vault API Request Details:
---[ REQUEST ]---------------------------------------
GET /v1/auth/token/lookup-self HTTP/1.1
Host: 127.0.0.1:8200
User-Agent: Go-http-client/1.1
X-Vault-Request: true
X-Vault-Token: hmac-sha256:0a983738db251df08f50dfed91f02effdfbcad8e0491bdd8da3c9e6be5b51315
Accept-Encoding: gzip


-----------------------------------------------------: timestamp="2024-11-05T03:14:31.701+0200"
2024-11-05T03:14:31.702+0200 [DEBUG] provider.terraform-provider-vault: [eb732054-0d99-45f1-99cd-9b0439a9a9a2] Vault API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Content-Length: 484
Cache-Control: no-store
Content-Type: application/json
Date: Tue, 05 Nov 2024 01:14:31 GMT
Strict-Transport-Security: max-age=31536000; includeSubDomains


-----------------------------------------------------: timestamp="2024-11-05T03:14:31.702+0200"
2024-11-05T03:14:31.702+0200 [DEBUG] provider.terraform-provider-vault: [e5f0a1f5-2563-46c1-a4bc-1fbe6e357976] Vault API Request Details:
---[ REQUEST ]---------------------------------------
POST /v1/auth/token/create HTTP/1.1
Host: 127.0.0.1:8200
User-Agent: Go-http-client/1.1
Content-Length: 128
X-Vault-Request: true
X-Vault-Token: hmac-sha256:48470e705aca7d5fa3894d993ded05aaf6a6701b6fca21d2268a2337a9ccd6a4
Accept-Encoding: gzip


-----------------------------------------------------: timestamp="2024-11-05T03:14:31.702+0200"
2024-11-05T03:14:31.703+0200 [DEBUG] provider.terraform-provider-vault: [e5f0a1f5-2563-46c1-a4bc-1fbe6e357976] Vault API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Content-Length: 455
Cache-Control: no-store
Content-Type: application/json
Date: Tue, 05 Nov 2024 01:14:31 GMT
Strict-Transport-Security: max-age=31536000; includeSubDomains


-----------------------------------------------------: timestamp="2024-11-05T03:14:31.703+0200"
2024-11-05T03:14:31.703+0200 [INFO]  provider.terraform-provider-vault: Using Vault token with the following policies: root: timestamp="2024-11-05T03:14:31.703+0200"
2024-11-05T03:14:31.703+0200 [DEBUG] provider.terraform-provider-vault: [6f286eb0-45d4-4350-b3d1-63ec2695293d] Vault API Request Details:
---[ REQUEST ]---------------------------------------
GET /v1/kv/data/test/acc456/env789/secret HTTP/1.1
Host: 127.0.0.1:8200
User-Agent: Go-http-client/1.1
X-Vault-Request: true
X-Vault-Token: hmac-sha256:867158e19d724fdf4977fcd6d83de8664e547f96b45def242683b086ca70a03e
Accept-Encoding: gzip


-----------------------------------------------------: timestamp="2024-11-05T03:14:31.703+0200"
2024-11-05T03:14:31.703+0200 [DEBUG] provider.terraform-provider-vault: [6f286eb0-45d4-4350-b3d1-63ec2695293d] Vault API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 404 Not Found
Content-Length: 14
Cache-Control: no-store
Content-Type: application/json
Date: Tue, 05 Nov 2024 01:14:31 GMT
Strict-Transport-Security: max-age=31536000; includeSubDomains


-----------------------------------------------------: timestamp="2024-11-05T03:14:31.703+0200"
2024-11-05T03:14:31.703+0200 [DEBUG] provider.terraform-provider-vault: Reading secret at "kv/data/test/acc456/env789/secret" (latest version) from Vault: timestamp="2024-11-05T03:14:31.703+0200"
2024-11-05T03:14:31.703+0200 [ERROR] provider.terraform-provider-vault: Response contains error diagnostic: tf_proto_version=5.4 @module=sdk.proto diagnostic_detail="" diagnostic_severity=ERROR diagnostic_summary="no secret found at \"kv/data/test/acc456/env789/secret\"" tf_data_source_type=vault_kv_secret_v2 tf_rpc=ReadDataSource @caller=/home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov5/internal/diag/diagnostics.go:62 tf_provider_addr=provider tf_req_id=9b5ff393-8f7d-c193-ecd7-e06c5857ed09 timestamp="2024-11-05T03:14:31.703+0200"
2024-11-05T03:14:31.703+0200 [ERROR] vertex "data.vault_kv_secret_v2.self" error: no secret found at "kv/data/test/acc456/env789/secret"
2024-11-05T03:14:31.703+0200 [ERROR] vertex "data.vault_kv_secret_v2.self (expand)" error: no secret found at "kv/data/test/acc456/env789/secret"
2024-11-05T03:14:31.704+0200 [DEBUG] TestFileRunner: completed plan for vault.tftest.hcl/validate_vault_secret
vault.tftest.hcl... fail
  run "validate_vault_secret"... fail
╷
│ Error: no secret found at "kv/data/test/acc456/env789/secret"
│
│   with data.vault_kv_secret_v2.self,
│   on main.tf line 29, in data "vault_kv_secret_v2" "self":
│   29: data "vault_kv_secret_v2" "self" {
│
╵

Failure! 0 passed, 1 failed.
2024-11-05T03:14:31.704+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-11-05T03:14:31.705+0200 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.opentofu.org/hashicorp/vault/4.4.0/darwin_arm64/terraform-provider-vault pid=54197
2024-11-05T03:14:31.705+0200 [DEBUG] provider: plugin exited
```

### Expected Behavior

When the vault is mocked, the address should not be mocked. If we need to define the vault provider with the address attribute, it will act as a duplicate declaration.

It should throw an error like below:
```zsh
╷
│ Error: Duplicate provider block
│
│   on vault.tftest.hcl line 16:
│   16: mock_provider "vault" {
│
│ A provider for vault.mock is already defined at vault.tftest.hcl:11,10-17.
```

The query should successfully retrieve the secret.

```zsh
vault.tftest.hcl... in progress
  run "validate_vault_secret"... pass

# data.vault_kv_secret_v2.self:
data "vault_kv_secret_v2" "self" {
    created_time    = "6q9mkyzt"
    custom_metadata = {}
    data            = (sensitive value)
    data_json       = (sensitive value)
    deletion_time   = "mdsx15xw"
    destroyed       = false
    id              = "okj5l3dg"
    mount           = "kv"
    name            = "test/acc456/env789/secret"
    path            = "2infs8mc"
}


Outputs:

vault_secret_data = (sensitive value)

vault.tftest.hcl... tearing down
vault.tftest.hcl... pass

Success! 1 passed, 0 failed.
```

### Actual Behavior

When vault is mocked, the address should be mocked as non-test provider, it will act as duplicate declaration. Also when we mention the `address` in the non test provider, it cannot get the mocked data from the mocked vault provider. 

Without non-test provider with address:
```zsh
$ tofu test
vault.tftest.hcl... fail
  run "validate_vault_secret"... fail
╷
│ Error: Missing required attribute
│
│   on provider["registry.opentofu.org/hashicorp/vault"] with no configuration line 1:
│   (source code not available)
│
│ The attribute "address" is required, but no definition was found.
╵

Failure! 0 passed, 1 failed.
```

With non-test provider with address:
```zsh
provider "vault" {
  alias   = "mock"
  address = "http://127.0.0.1:8200"
}
```

```zsh
vault.tftest.hcl... fail
  run "validate_vault_secret"... fail
╷
│ Error: no secret found at "kv/data/test/acc456/env789/secret"
│
│   with data.vault_kv_secret_v2.self,
│   on main.tf line 29, in data "vault_kv_secret_v2" "self":
│   29: data "vault_kv_secret_v2" "self" {
│
╵

Failure! 0 passed, 1 failed.
```

### Steps to Reproduce

1. `tofu init`
2. `tofu test`

### Additional Context

This occurs when we perform the `tofu test`. I used Terraform version 1.9.1, which does not require the address attribute declaration, and the secret is picked up correctly.

### References

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -56,6 +56,7 @@ BUG FIXES:
 * Error message about a provider type mismatch now correctly identifies which module contains the problem. ([#1991](https://github.com/opentofu/opentofu/pull/1991))
 * The `yamldecode` function's interpretation of scalars as numbers now conforms to the YAML 1.2 specification. In particular, the scalar value `+` is now interpreted as the string `"+"` rather than returning a parse error trying to interpret it as an integer. ([#2044](https://github.com/opentofu/opentofu/pull/2044))
 * A `module` block's `version` argument now accepts prerelease version selections using a "v" prefix before the version number. Previously this was accepted only for non-prerelease selections. ([#2124])(https://github.com/opentofu/opentofu/issues/2124)
+* The `tofu test` command doesn't try to validate mock provider definition by its underlying provider schema now. ([#2140](https://github.com/opentofu/opentofu/pull/2140))
 
 INTERNAL CHANGES:
 

--- a/internal/command/test_test.go
+++ b/internal/command/test_test.go
@@ -17,6 +17,7 @@ import (
 	"github.com/opentofu/opentofu/internal/addrs"
 	testing_command "github.com/opentofu/opentofu/internal/command/testing"
 	"github.com/opentofu/opentofu/internal/command/views"
+	"github.com/opentofu/opentofu/internal/configs/configschema"
 	"github.com/opentofu/opentofu/internal/providers"
 	"github.com/opentofu/opentofu/internal/terminal"
 )
@@ -1394,5 +1395,68 @@ digits, underscores, and dashes.
 				t.Fatalf("expected status code 0 but got %d: %s", code, ui.ErrorWriter)
 			}
 		})
+	}
+}
+
+// TestTest_MockProviderValidation checks if tofu test runs proper validation for
+// mock_provider. Even if provider schema has required fields, tofu test should
+// ignore it completely, because the provider is mocked.
+func TestTest_MockProviderValidation(t *testing.T) {
+	td := t.TempDir()
+	testCopyDir(t, testFixturePath("test/mock_provider_validation"), td)
+	defer testChdir(t, td)()
+
+	provider := testing_command.NewProvider(nil)
+	providerSource, close := newMockProviderSource(t, map[string][]string{
+		"test": {"1.0.0"},
+	})
+	defer close()
+
+	provider.Provider.GetProviderSchemaResponse = &providers.GetProviderSchemaResponse{
+		Provider: providers.Schema{
+			Block: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"required_field": {
+						Type:     cty.String,
+						Required: true,
+					},
+				},
+			},
+		},
+		ResourceTypes: map[string]providers.Schema{
+			"test_resource": {
+				Block: &configschema.Block{
+					Attributes: map[string]*configschema.Attribute{
+						"value": {
+							Type:     cty.String,
+							Optional: true,
+						},
+						"computed_value": {
+							Type:     cty.String,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	streams, _ := terminal.StreamsForTesting(t)
+	view := views.NewView(streams)
+	ui := new(cli.MockUi)
+	meta := Meta{
+		testingOverrides: metaOverridesForProvider(provider.Provider),
+		Ui:               ui,
+		View:             view,
+		Streams:          streams,
+		ProviderSource:   providerSource,
+	}
+
+	testCmd := &TestCommand{
+		Meta: meta,
+	}
+
+	if code := testCmd.Run(nil); code != 0 {
+		t.Fatalf("expected status code 0 but got %d: %s", code, ui.ErrorWriter)
 	}
 }

--- a/internal/command/test_test.go
+++ b/internal/command/test_test.go
@@ -1407,10 +1407,10 @@ func TestTest_MockProviderValidation(t *testing.T) {
 	defer testChdir(t, td)()
 
 	provider := testing_command.NewProvider(nil)
-	providerSource, close := newMockProviderSource(t, map[string][]string{
+	providerSource, closePS := newMockProviderSource(t, map[string][]string{
 		"test": {"1.0.0"},
 	})
-	defer close()
+	defer closePS()
 
 	provider.Provider.GetProviderSchemaResponse = &providers.GetProviderSchemaResponse{
 		Provider: providers.Schema{

--- a/internal/command/testdata/test/mock_provider_validation/main.tf
+++ b/internal/command/testdata/test/mock_provider_validation/main.tf
@@ -1,0 +1,5 @@
+provider "test" {}
+
+resource "test_resource" "primary" {
+  value = "foo"
+}

--- a/internal/command/testdata/test/mock_provider_validation/main.tftest.hcl
+++ b/internal/command/testdata/test/mock_provider_validation/main.tftest.hcl
@@ -1,0 +1,14 @@
+mock_provider "test" {
+  mock_resource "test_resource" {
+    defaults = {
+      computed_value = "bar"      
+    }
+  }
+}
+
+run "test" {
+  assert {
+    condition = test_resource.primary.computed_value == "bar"
+    error_message = "Unexpected computed value"
+  }
+}

--- a/internal/tofu/provider_for_test_framework.go
+++ b/internal/tofu/provider_for_test_framework.go
@@ -104,7 +104,7 @@ func (p *providerForTest) ReadDataSource(r providers.ReadDataSourceRequest) prov
 }
 
 // ValidateProviderConfig is irrelevant when provider is mocked or overridden.
-func (p *providerForTest) ValidateProviderConfig(r providers.ValidateProviderConfigRequest) providers.ValidateProviderConfigResponse {
+func (p *providerForTest) ValidateProviderConfig(_ providers.ValidateProviderConfigRequest) providers.ValidateProviderConfigResponse {
 	return providers.ValidateProviderConfigResponse{}
 }
 

--- a/internal/tofu/provider_for_test_framework.go
+++ b/internal/tofu/provider_for_test_framework.go
@@ -110,7 +110,10 @@ func (p *providerForTest) ValidateProviderConfig(_ providers.ValidateProviderCon
 
 // GetProviderSchema is also used to perform additional validation outside of the provider
 // implementation. We are excluding parts of the schema related to provider since it is
-// irrelevant in the scope of mocking / overriding.
+// irrelevant in the scope of mocking / overriding. When running `tofu test` configuration
+// is being transformed for testing framework and original provider configuration is not
+// accessible so it is safe to wipe metadata as well. See Config.transformProviderConfigsForTest
+// for more details.
 func (p *providerForTest) GetProviderSchema() providers.GetProviderSchemaResponse {
 	providerSchema := p.internal.GetProviderSchema()
 	providerSchema.Provider = providers.Schema{}

--- a/internal/tofu/provider_for_test_framework.go
+++ b/internal/tofu/provider_for_test_framework.go
@@ -118,41 +118,9 @@ func (p *providerForTest) GetProviderSchema() providers.GetProviderSchemaRespons
 	return providerSchema
 }
 
-// Calling the internal provider ensures providerForTest has the same behaviour as if
-// it wasn't overridden or mocked. The only exception is ImportResourceState, which panics
-// if called via providerForTest because importing is not supported in testing framework.
-
-func (p *providerForTest) ValidateResourceConfig(r providers.ValidateResourceConfigRequest) providers.ValidateResourceConfigResponse {
-	return p.internal.ValidateResourceConfig(r)
-}
-
-func (p *providerForTest) ValidateDataResourceConfig(r providers.ValidateDataResourceConfigRequest) providers.ValidateDataResourceConfigResponse {
-	return p.internal.ValidateDataResourceConfig(r)
-}
-
-func (p *providerForTest) UpgradeResourceState(r providers.UpgradeResourceStateRequest) providers.UpgradeResourceStateResponse {
-	return p.internal.UpgradeResourceState(r)
-}
-
 // providerForTest doesn't configure its internal provider because it is mocked.
 func (p *providerForTest) ConfigureProvider(_ providers.ConfigureProviderRequest) providers.ConfigureProviderResponse {
 	return providers.ConfigureProviderResponse{}
-}
-
-func (p *providerForTest) Stop() error {
-	return p.internal.Stop()
-}
-
-func (p *providerForTest) GetFunctions() providers.GetFunctionsResponse {
-	return p.internal.GetFunctions()
-}
-
-func (p *providerForTest) CallFunction(r providers.CallFunctionRequest) providers.CallFunctionResponse {
-	return p.internal.CallFunction(r)
-}
-
-func (p *providerForTest) Close() error {
-	return p.internal.Close()
 }
 
 func (p *providerForTest) ImportResourceState(providers.ImportResourceStateRequest) providers.ImportResourceStateResponse {
@@ -195,6 +163,38 @@ func (p *providerForTest) addMockResources(mockResources []*configs.MockResource
 			overrideValues: mockRes.Defaults,
 		}
 	}
+}
+
+// Calling the internal provider ensures providerForTest has the same behaviour as if
+// it wasn't overridden or mocked. The only exception is ImportResourceState, which panics
+// if called via providerForTest because importing is not supported in testing framework.
+
+func (p *providerForTest) ValidateResourceConfig(r providers.ValidateResourceConfigRequest) providers.ValidateResourceConfigResponse {
+	return p.internal.ValidateResourceConfig(r)
+}
+
+func (p *providerForTest) ValidateDataResourceConfig(r providers.ValidateDataResourceConfigRequest) providers.ValidateDataResourceConfigResponse {
+	return p.internal.ValidateDataResourceConfig(r)
+}
+
+func (p *providerForTest) UpgradeResourceState(r providers.UpgradeResourceStateRequest) providers.UpgradeResourceStateResponse {
+	return p.internal.UpgradeResourceState(r)
+}
+
+func (p *providerForTest) Stop() error {
+	return p.internal.Stop()
+}
+
+func (p *providerForTest) GetFunctions() providers.GetFunctionsResponse {
+	return p.internal.GetFunctions()
+}
+
+func (p *providerForTest) CallFunction(r providers.CallFunctionRequest) providers.CallFunctionResponse {
+	return p.internal.CallFunction(r)
+}
+
+func (p *providerForTest) Close() error {
+	return p.internal.Close()
 }
 
 type resourceForTest struct {

--- a/internal/tofu/provider_for_test_framework.go
+++ b/internal/tofu/provider_for_test_framework.go
@@ -103,17 +103,24 @@ func (p *providerForTest) ReadDataSource(r providers.ReadDataSourceRequest) prov
 	return resp
 }
 
+// ValidateProviderConfig is irrelevant when provider is mocked or overridden.
+func (p *providerForTest) ValidateProviderConfig(r providers.ValidateProviderConfigRequest) providers.ValidateProviderConfigResponse {
+	return providers.ValidateProviderConfigResponse{}
+}
+
+// GetProviderSchema is also used to perform additional validation outside of the provider
+// implementation. We are excluding parts of the schema related to provider since it is
+// irrelevant in the scope of mocking / overriding.
+func (p *providerForTest) GetProviderSchema() providers.GetProviderSchemaResponse {
+	providerSchema := p.internal.GetProviderSchema()
+	providerSchema.Provider = providers.Schema{}
+	providerSchema.ProviderMeta = providers.Schema{}
+	return providerSchema
+}
+
 // Calling the internal provider ensures providerForTest has the same behaviour as if
 // it wasn't overridden or mocked. The only exception is ImportResourceState, which panics
 // if called via providerForTest because importing is not supported in testing framework.
-
-func (p *providerForTest) GetProviderSchema() providers.GetProviderSchemaResponse {
-	return p.internal.GetProviderSchema()
-}
-
-func (p *providerForTest) ValidateProviderConfig(r providers.ValidateProviderConfigRequest) providers.ValidateProviderConfigResponse {
-	return p.internal.ValidateProviderConfig(r)
-}
 
 func (p *providerForTest) ValidateResourceConfig(r providers.ValidateResourceConfigRequest) providers.ValidateResourceConfigResponse {
 	return p.internal.ValidateResourceConfig(r)

